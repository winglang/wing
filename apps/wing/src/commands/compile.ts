import { promises as fsPromise } from "fs";
import { dirname, relative, resolve } from "path";

import * as wingCompiler from "@winglang/compiler";
import { prettyPrintError } from "@winglang/sdk/lib/util/enhanced-error";
import chalk from "chalk";
import { CHARS_ASCII, emitDiagnostic, File, Label } from "codespan-wasm";
import debug from "debug";
import { glob } from "glob";
import { loadEnvVariables } from "../env";

// increase the stack trace limit to 50, useful for debugging Rust panics
// (not setting the limit too high in case of infinite recursion)
Error.stackTraceLimit = 50;

const log = debug("wing:compile");

/**
 * Compile options for the `compile` command.
 * This is passed from Commander to the `compile` function.
 */
export interface CompileOptions {
  /**
   * Target platform
   * @default wingCompiler.BuiltinPlatform.SIM
   */
  readonly platform: string[];
  /**
   * App root id
   *
   * @default "Default"
   */
  readonly rootId?: string;
  /**
   * String with platform-specific values separated by commas
   */
  readonly value?: string;
  /**
   * Path to the file with specific platform values (TOML|YAML|JSON)
   *
   * example of the file's content:
   * root/Default/Domain:
   *   hostedZoneId: Z0111111111111111111F
   *   acmCertificateArn: arn:aws:acm:us-east-1:111111111111:certificate/aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee
   */
  readonly values?: string;
  /**
   * The location to save the compilation output
   * @default "./target"
   */
  readonly targetDir?: string;
  /**
   * The overrides the location to save the compilation output
   * @default "./target/<entrypoint>.<target>"
   */
  readonly output?: string;
  /**
   * Whether to run the compiler in `wing test` mode. This may create multiple
   * copies of the application resources in order to run tests in parallel.
   */
  readonly testing?: boolean;
}

/**
 * Compiles a Wing program. Throws an error if compilation fails.
 * @param entrypoint The program .w entrypoint.
 * @param options Compile options.
 * @returns the output directory
 */
export async function compile(entrypoint?: string, options?: CompileOptions): Promise<string> {
  if (!entrypoint) {
    const wingFiles = (await glob("{main,*.main}.{w,ts}")).sort();
    if (wingFiles.length === 0) {
      throw new Error(
        "Cannot find an entrypoint file (main.w, main.ts, *.main.w, *.main.ts) in the current directory."
      );
    }
    if (wingFiles.length > 1) {
      throw new Error(
        `Multiple entrypoints found in the current directory (${wingFiles.join(
          ", "
        )}). Please specify which one to use.`
      );
    }
    entrypoint = wingFiles[0];
  }
  loadEnvVariables({ cwd: resolve(dirname(entrypoint)) });
  const coloring = chalk.supportsColor ? chalk.supportsColor.hasBasic : false;
  try {
    return await wingCompiler.compile(entrypoint, {
      ...options,
      log,
      color: coloring,
      platform: options?.platform ?? ["sim"],
    });
  } catch (error) {
    if (error instanceof wingCompiler.CompileError) {
      // This is a bug in the user's code. Print the compiler diagnostics.
      const diagnostics = error.diagnostics;
      const cwd = process.cwd();
      const result = [];

      for (const diagnostic of diagnostics) {
        const { message, span, annotations, hints } = diagnostic;
        const files: File[] = [];
        const labels: Label[] = [];

        // file_id might be "" if the span is synthetic (see #2521)
        if (span?.file_id) {
          // `span` should only be null if source file couldn't be read etc.
          const source = await fsPromise.readFile(span.file_id, "utf8");
          const start = span.start_offset;
          const end = span.end_offset;
          const filePath = relative(cwd, span.file_id);
          files.push({ name: filePath, source });
          labels.push({
            fileId: filePath,
            rangeStart: start,
            rangeEnd: end,
            message: "",
            style: "primary",
          });
        }

        for (const annotation of annotations) {
          // file_id might be "" if the span is synthetic (see #2521)
          if (!annotation.span?.file_id) {
            continue;
          }
          const source = await fsPromise.readFile(annotation.span.file_id, "utf8");
          const start = annotation.span.start_offset;
          const end = annotation.span.end_offset;
          const filePath = relative(cwd, annotation.span.file_id);
          files.push({ name: filePath, source });
          labels.push({
            fileId: filePath,
            rangeStart: start,
            rangeEnd: end,
            message: annotation.message,
            style: "secondary",
          });
        }

        const diagnosticText = emitDiagnostic(
          files,
          {
            message,
            severity: "error",
            labels,
            notes: hints.map((hint) => `hint: ${hint}`),
          },
          {
            chars: CHARS_ASCII,
          },
          coloring
        );
        result.push(diagnosticText);
      }
      throw new Error(result.join("\n"));
    } else if (error instanceof wingCompiler.PreflightError) {
      let output = await prettyPrintError(error.causedBy, {
        chalk,
        sourceEntrypoint: resolve(entrypoint ?? "."),
      });

      if (process.env.DEBUG) {
        output +=
          "\n--------------------------------- ORIGINAL STACK TRACE ---------------------------------\n" +
          (error.causedBy.stack ?? "(no stacktrace available)");
      }

      error.causedBy.message = output;

      throw error.causedBy;
    } else {
      throw error;
    }
  }
}
