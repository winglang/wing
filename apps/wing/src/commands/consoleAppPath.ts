export async function consoleAppPath() {
  const path = require.resolve("@wingconsole/app");
  console.log(path);
}
