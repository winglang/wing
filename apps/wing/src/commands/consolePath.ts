export async function consolePath() {
  const wingconsolePath = require.resolve("@wingconsole/app");
  console.log(wingconsolePath);
}
