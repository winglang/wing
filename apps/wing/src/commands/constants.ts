/**
 * Available targets for compilation.
 * This is passed from Commander to the `compile` function.
 */
export enum Target {
  TF_AWS = "tf-aws",
  TF_AZURE = "tf-azure",
  TF_GCP = "tf-gcp",
  SIM = "sim",
  AWSCDK = "awscdk",
}
