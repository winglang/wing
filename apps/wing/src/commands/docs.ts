import open from "open";

export async function docs() {
  await open("https://docs.winglang.io");
}