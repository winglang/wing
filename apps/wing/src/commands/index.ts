export * from "./compile";
export * from "./upgrade";
export * from "./test";
export * from "./docs";
export * from "./run";
