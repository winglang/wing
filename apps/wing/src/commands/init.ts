// This file and the main function are named "init" instead of "new"
// to avoid a conflict with the "new" keyword in JavaScript
import { exec } from "child_process";
import { existsSync } from "fs";
import { copyFile, mkdir, readFile, readdir, writeFile } from "fs/promises";
import { join, relative } from "path";
import { promisify } from "util";
import chalk from "chalk";
import inquirer from "inquirer";
import { exists } from "./pack";
import { projectTemplateNames, PROJECT_TEMPLATES_DIR, currentPackage } from "../util";

const execPromise = promisify(exec);

/**
 * Options for the `new` command.
 * This is passed from Commander to the `init` function.
 */
export interface InitOptions {
  readonly template?: string;
  readonly language?: string;
  readonly listTemplates?: boolean;
}

/**
 * Compiles a Wing program. Throws an error if compilation fails.
 * @param entrypoint The program .w entrypoint.
 * @param options Compile options.
 * @returns the output directory
 */
export async function init(template: string, options: InitOptions = {}): Promise<void> {
  const templates = projectTemplateNames();
  let language = options.language ?? "wing";

  // If --list-templates is specified, list the available templates and exit
  if (options.listTemplates) {
    console.log(templates.join("\n"));
    return;
  }

  // If no template is specified, let them interactively select one
  if (!template) {
    console.log("Usage: wing new <template> [--language <language>]");
    console.log();
    try {
      const responses = await inquirer.prompt([
        {
          type: "list",
          name: "template",
          message: "Please select a template:",
          choices: templates,
        },
        {
          type: "list",
          name: "language",
          message: "Please select a language:",
          choices: ["wing", "typescript"],
        },
      ]);
      template = responses.template;
      language = responses.language;
      console.log();
    } catch (err) {
      if ((err as any).isTtyError) {
        throw new Error(
          `Please select from one of the available choices:\n  ${templates.join(
            "\n  "
          )}\n\nHave an idea for a new template? Let us know at https://github.com/winglang/wing/issues/!`
        );
      }
    }
  }

  if (!templates.includes(template)) {
    throw new Error(
      `Template "${template}" is not available. Please select from one of the available choices:\n  ${templates.join(
        "\n  "
      )}\n\nHave an idea for a new template? Let us know at https://github.com/winglang/wing/issues/!`
    );
  }

  // Parse the language selected
  switch (language) {
    case "ts":
    case "typescript":
      language = "typescript";
      break;
    case "wing":
    case "winglang":
      language = "wing";
      break;
    default:
      throw new Error(
        `Unknown language: "${language}". Please select from "wing" or "typescript".`
      );
  }

  // Check if the template exists for the selected language
  const templatePath = join(PROJECT_TEMPLATES_DIR, language, template);
  const templateExists = await exists(templatePath);
  if (!templateExists) {
    throw new Error(
      `Template "${template}" is not available in ${language}. Please let us know you'd like to use this template in ${language} by opening an issue at https://github.com/winglang/wing/issues/!`
    );
  }

  // Obtain a list of files in the template, and check they won't overwrite anything
  const files = await getFiles(templatePath);
  const existingFiles = await getFiles(process.cwd());
  const overwrite = existingFiles.filter((file) => files.includes(file));
  if (overwrite.length > 0) {
    throw new Error(
      `The following files already exist in the current directory and will be overwritten:\n  ${overwrite.join(
        "\n  "
      )}\n\nPlease move or delete these files and try again.`
    );
  }

  // Copy the template
  await copyFiles(templatePath, process.cwd());

  // Replace wing version
  const packageJsonPath = join(process.cwd(), "package.json");
  if (existsSync(packageJsonPath)) {
    const packageJson = JSON.parse(await readFile(packageJsonPath, "utf8"));
    const isDevVersion = currentPackage.version.startsWith("0.0.0");

    const depKeys = ["dependencies", "devDependencies", "peerDependencies"];
    for (const depKey of depKeys) {
      const depMap = packageJson[depKey] ?? {};
      for (const [key, val] of Object.entries(depMap)) {
        if ((val as string).includes("#WING_VERSION#")) {
          if (isDevVersion) {
            if (key === "winglang") {
              depMap[key] = `file:${join(__dirname, "..", "..")}`;
            } else {
              depMap[key] = `file:${join(__dirname, "..", "..", "..", "..", "libs", key)}`;
            }
          } else {
            depMap[key] = depMap[key].replace("#WING_VERSION#", currentPackage.version);
          }
        }
      }
    }

    await writeFile(packageJsonPath, JSON.stringify(packageJson, null, 2));
  }

  // Run npm install
  console.log("Installing dependencies...");
  console.log();

  // Check if npm is installed
  const npmExists = await execPromise("npm --version")
    .then(() => true)
    .catch(() => false);
  if (!npmExists) {
    console.log(
      `${chalk.yellow(
        "warning:"
      )} npm is not installed. Please install npm and run "npm install" to finish setting up any project dependencies.`
    );
  }

  // Install dependencies (in the current directory)
  try {
    await execPromise("npm install");
  } catch (err) {
    console.log(
      `${chalk.yellow(
        "warning:"
      )} npm install failed. Please let us know there's an issue with this template by opening an issue at at https://github.com/winglang/wing/issues/.`
    );
    console.log();
    console.error((err as any).stderr);
  }

  console.log(`Created a new ${chalk.cyan(template)} project in the current directory! ðŸŽ‰`);
  console.log();
  console.log("Not sure where to get started? Try running:");
  console.log();
  console.log("  wing compile - build your project");
  console.log("  wing it - simulate your app in the Wing Console");
  console.log("  wing test - run all tests");
  console.log();
  console.log("Visit the docs for examples and tutorials: https://winglang.io/docs");
}

/**
 * Recursively copies a directory.
 * @param src The source directory.
 * @param dest The destination directory.
 */
async function copyFiles(src: string, dest: string): Promise<void> {
  // Create the destination directory if it doesn't exist
  if (!existsSync(dest)) {
    await mkdir(dest);
  }

  // Copy all files
  const files = await getFiles(src);
  for (const file of files) {
    await copyFile(join(src, file), join(dest, file));
  }
}

/**
 * Recursively gets all files in a directory.
 * @param dir The directory to search.
 * @returns a list of files.
 */
async function getFiles(dir: string): Promise<string[]> {
  return getFilesHelper(dir, dir);
}

async function getFilesHelper(basedir: string, dir: string): Promise<string[]> {
  const dirents = await readdir(dir, { withFileTypes: true });
  const files = await Promise.all(
    dirents.map(async (dirent) => {
      const res = join(dir, dirent.name);
      return dirent.isDirectory() ? getFilesHelper(basedir, res) : relative(basedir, res);
    })
  );
  return Array.prototype.concat(...files);
}
