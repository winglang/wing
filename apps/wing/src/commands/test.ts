import { basename, resolve, sep } from "path";
import { compile, CompileOptions } from "./compile";
import chalk from "chalk";
import { std, simulator } from "@winglang/sdk";
import * as cp from "child_process";
import debug from "debug";
import { promisify } from "util";
import { generateTmpDir, withSpinner } from "../util";
import { Target } from "@winglang/compiler";
import { nanoid } from "nanoid";
import { readFile, rm, rmSync } from "fs";

const log = debug("wing:test");

const ENV_WING_TEST_RUNNER_FUNCTION_ARNS = "WING_TEST_RUNNER_FUNCTION_ARNS";
const ENV_WING_TEST_RUNNER_FUNCTION_ARNS_AWSCDK = "WingTestRunnerFunctionArns";

/**
 * @param path path to the test/s file
 * @returns the file name and parent dir in the following format: "folder/file.ext"
 */
const generateTestName = (path: string) => path.split(sep).slice(-2).join("/");

/**
 * Options for the `test` command.
 */
export interface TestOptions extends CompileOptions {
  clean: boolean;
}

export async function test(entrypoints: string[], options: TestOptions): Promise<number> {
  const startTime = Date.now();
  const results: { testName: string; results: std.TestResult[] }[] = [];
  const testFile = async (entrypoint: string) => {
    const testName = generateTestName(entrypoint);
    try {
      const singleTestResults: std.TestResult[] | void = await testOne(entrypoint, options);
      results.push({ testName, results: singleTestResults ?? [] });
    } catch (error) {
      console.log((error as Error).message);
      results.push({
        testName: generateTestName(entrypoint),
        results: [{ pass: false, path: "", error: (error as Error).message, traces: [] }],
      });
    }
  };
  await Promise.all(entrypoints.map(testFile));
  printResults(results, Date.now() - startTime);

  // if we have any failures, exit with 1
  for (const test of results) {
    for (const r of test.results) {
      if (r.error) {
        return 1;
      }
    }
  }

  return 0;
}

function printResults(
  testResults: { testName: string; results: std.TestResult[] }[],
  duration: number
) {
  const durationInSeconds = duration / 1000;
  const totalSum = testResults.length;
  const failing = testResults.filter(({ results }) => results.some(({ pass }) => !pass));
  const passing = testResults.filter(({ results }) => results.every(({ pass }) => !!pass));
  const failingTestsNumber = failing.reduce(
    (acc, { results }) => acc + results.filter(({ pass }) => !pass).length,
    0
  );
  const passingTestsNumber = testResults.reduce(
    (acc, { results }) => acc + results.filter(({ pass }) => !!pass).length,
    0
  );
  console.log(" "); // for getting a new line- \n does't seem to work :(
  const areErrors = failing.length > 0 && totalSum > 1;
  const showTitle = totalSum > 1;

  const results = [];

  if (showTitle) {
    // prints a list of the tests names with an icon
    results.push(`Results:`);
    results.push(...passing.map(({ testName }) => `    ${chalk.green("✓")} ${testName}`));
    results.push(...failing.map(({ testName }) => `    ${chalk.red("×")} ${testName}`));
  }

  if (areErrors) {
    // prints error messages form failed tests
    results.push(" ");
    results.push("Errors:");
    results.push(
      ...failing.map(({ testName, results }) =>
        [
          `At ${testName}`,
          results.filter(({ pass }) => !pass).map(({ error }) => chalk.red(error)),
        ].join("\n")
      )
    );
  }

  // prints a summary of how many tests passed and failed
  results.push(" ");
  results.push(
    `${chalk.dim("Tests")}${failingTestsNumber ? chalk.red(` ${failingTestsNumber} failed`) : ""}${
      failingTestsNumber && passingTestsNumber ? chalk.dim(" |") : ""
    }${passingTestsNumber ? chalk.green(` ${passingTestsNumber} passed`) : ""} ${chalk.dim(
      `(${failingTestsNumber + passingTestsNumber})`
    )}`
  );
  // prints a summary of how many tests files passed and failed
  results.push(
    `${chalk.dim("Test Files")}${failing.length ? chalk.red(` ${failing.length} failed`) : ""}${
      failing.length && passing.length ? chalk.dim(" |") : ""
    }${passing.length ? chalk.green(` ${passing.length} passed`) : ""} ${chalk.dim(
      `(${totalSum})`
    )}`
  );

  // prints the test duration
  results.push(
    `${chalk.dim("Duration")} ${Math.floor(durationInSeconds / 60)}m${(
      durationInSeconds % 60
    ).toFixed(2)}s`
  );

  console.log(results.filter((value) => !!value).join("\n"));
}

async function testOne(entrypoint: string, options: TestOptions) {
  const synthDir = await withSpinner(
    `Compiling ${generateTestName(entrypoint)} to ${options.target}...`,
    async () =>
      compile(entrypoint, {
        ...options,
        rootId: options.rootId ?? `Test.${nanoid(10)}`,
        testing: true,
        // since the test cleans up after each run, it's essential to create a temporary output directory-
        // at least one that is different then the usual compilation output dir,  otherwise we might end up cleaning up the user's actual resources.
        ...(options.target !== Target.SIM && { targetDir: `${await generateTmpDir()}/target` }),
      })
  );

  switch (options.target) {
    case Target.SIM:
      return await testSimulator(synthDir, options);
    case Target.TF_AWS:
      return await testTfAws(synthDir, options);
    case Target.AWSCDK:
      return await testAwsCdk(synthDir, options);
    default:
      throw new Error(`unsupported target ${options.target}`);
  }
}

/**
 * Render a test report for printing out to the console.
 */
export function renderTestReport(entrypoint: string, results: std.TestResult[]): string {
  const out = new Array<string>();

  // find the longest `path` of all the tests
  const longestPath = results.reduce(
    (longest, result) => (result.path.length > longest ? result.path.length : longest),
    0
  );

  // if there are no inflight tests, add a dummy "pass" result
  // to indicate that compilation and preflight checks passed
  if (results.length === 0) {
    results.push({
      pass: true,
      path: "",
      traces: [],
    });
  }

  for (const result of results.sort(sortTests)) {
    const status = result.pass ? chalk.green("pass") : chalk.red("fail");

    const details = new Array<string>();

    // add any log messages that were emitted during the test
    for (const log of result.traces) {
      // only show detailed traces if we are in debug mode
      if (log.type === "resource" && process.env.DEBUG) {
        details.push(chalk.gray("[trace] " + log.data.message));
      }
      if (log.type === "log") {
        details.push(chalk.gray(log.data.message));
      }
    }

    // if the test failed, add the error message and trace
    if (result.error) {
      details.push(...result.error.split("\n").map((l) => chalk.red(l)));
    }

    // construct the first row of the test result by collecting the various components and joining
    // them with spaces.

    const firstRow = new Array<string>();
    firstRow.push(status);

    // if we have details, surround the rows with a box, otherwise, just print a line
    if (details.length > 0) {
      firstRow.push(chalk.gray("┌"));
    } else {
      firstRow.push(chalk.gray("─"));
    }

    firstRow.push(basename(entrypoint));

    if (result.path?.length > 0) {
      firstRow.push(chalk.gray("»"));
      firstRow.push(chalk.whiteBright(result.path.padEnd(longestPath)));
    } else {
      firstRow.push(chalk.gray("(no tests)"));
    }

    // okay we are ready to print the test result

    // print the primary description of the test
    out.push(firstRow.join(" "));

    // print additional rows that are related to this test
    for (let i = 0; i < details.length; i++) {
      const left = i === details.length - 1 ? "└" : "│";
      out.push(`    ${chalk.gray(` ${left} `)}${details[i]}`);
    }
  }

  return out.join("\n");
}

function testResultsContainsFailure(results: std.TestResult[]): boolean {
  return results.some((r) => !r.pass);
}

function noCleanUp(synthDir: string) {
  console.log(
    chalk.yellowBright.bold(`Cleanup is disabled!\nOutput files available at ${resolve(synthDir)}`)
  );
}

async function testSimulator(synthDir: string, options: TestOptions) {
  const s = new simulator.Simulator({ simfile: synthDir });
  const { clean } = options;
  await s.start();

  const testRunner = s.getResource("root/cloud.TestRunner") as std.ITestRunnerClient;
  const tests = await testRunner.listTests();
  const filteredTests = pickOneTestPerEnvironment(tests);
  const results = new Array<std.TestResult>();

  // TODO: run these tests in parallel
  for (const path of filteredTests) {
    results.push(await testRunner.runTest(path));
  }

  await s.stop();

  const testReport = renderTestReport(synthDir, results);
  console.log(testReport);

  if (clean) {
    rmSync(synthDir, { recursive: true, force: true });
  } else {
    noCleanUp(synthDir);
  }

  return results;
}

async function testAwsCdk(synthDir: string, options: TestOptions): Promise<std.TestResult[]> {
  const { clean } = options;
  try {
    isAwsCdkInstalled(synthDir);

    await withSpinner("cdk deploy", () => awsCdkDeploy(synthDir));

    const [testRunner, tests] = await withSpinner("Setting up test runner...", async () => {
      const testArns = await awsCdkOutput(
        synthDir,
        ENV_WING_TEST_RUNNER_FUNCTION_ARNS_AWSCDK,
        process.env.CDK_STACK_NAME!
      );

      const { TestRunnerClient } = await import(
        "@winglang/sdk/lib/shared-aws/test-runner.inflight"
      );
      const testRunner = new TestRunnerClient(testArns);

      const tests = await testRunner.listTests();
      return [testRunner, pickOneTestPerEnvironment(tests)];
    });

    const results = await withSpinner("Running tests...", async () => {
      const results = new Array<std.TestResult>();
      for (const path of tests) {
        results.push(await testRunner.runTest(path));
      }
      return results;
    });

    const testReport = renderTestReport(synthDir, results);
    console.log(testReport);

    if (testResultsContainsFailure(results)) {
      console.log("One or more tests failed. Cleaning up resources...");
    }

    return results;
  } catch (err) {
    console.warn((err as Error).message);
    return [{ pass: false, path: "", error: (err as Error).message, traces: [] }];
  } finally {
    if (clean) {
      await cleanupCdk(synthDir);
    } else {
      noCleanUp(synthDir);
    }
  }
}

async function cleanupCdk(synthDir: string) {
  await withSpinner("aws-cdk destroy", () => awsCdkDestroy(synthDir));
  rmSync(synthDir, { recursive: true, force: true });
}

async function isAwsCdkInstalled(synthDir: string) {
  try {
    await execCapture("cdk version --ci true", { cwd: synthDir });
  } catch (err) {
    throw new Error(
      "AWS-CDK is not installed. Please install AWS-CDK to run tests in the cloud (npm i -g aws-cdk)."
    );
  }
}

export async function awsCdkDeploy(synthDir: string) {
  await execCapture("cdk deploy --require-approval never --ci true -O ./output.json --app . ", {
    cwd: synthDir,
  });
}

export async function awsCdkDestroy(synthDir: string) {
  const removeFile = promisify(rm);
  await removeFile(synthDir.concat("/output.json"));
  await execCapture("cdk destroy -f --ci true --app ./", { cwd: synthDir });
}

async function awsCdkOutput(synthDir: string, name: string, stackName: string) {
  const readFileCmd = promisify(readFile);
  const file = await readFileCmd(synthDir.concat("/output.json"));
  const parsed = JSON.parse(Buffer.from(file).toString());
  return parsed[stackName][name];
}

async function testTfAws(synthDir: string, options: TestOptions): Promise<std.TestResult[] | void> {
  const { clean } = options;
  try {
    if (!isTerraformInstalled(synthDir)) {
      throw new Error(
        "Terraform is not installed. Please install Terraform to run tests in the cloud."
      );
    }

    await withSpinner("terraform init", async () => await terraformInit(synthDir));

    await withSpinner("terraform apply", () => terraformApply(synthDir));

    const [testRunner, tests] = await withSpinner("Setting up test runner...", async () => {
      const testArns = await terraformOutput(synthDir, ENV_WING_TEST_RUNNER_FUNCTION_ARNS);
      const { TestRunnerClient } = await import(
        "@winglang/sdk/lib/shared-aws/test-runner.inflight"
      );
      const testRunner = new TestRunnerClient(testArns);

      const tests = await testRunner.listTests();
      return [testRunner, pickOneTestPerEnvironment(tests)];
    });

    const results = await withSpinner("Running tests...", async () => {
      const results = new Array<std.TestResult>();
      for (const path of tests) {
        results.push(await testRunner.runTest(path));
      }
      return results;
    });

    const testReport = renderTestReport(synthDir, results);
    console.log(testReport);

    if (testResultsContainsFailure(results)) {
      console.log("One or more tests failed. Cleaning up resources...");
    }

    return results;
  } catch (err) {
    console.warn((err as Error).message);
    return [{ pass: false, path: "", error: (err as Error).message, traces: [] }];
  } finally {
    if (clean) {
      await cleanupTf(synthDir);
    } else {
      noCleanUp(synthDir);
    }
  }
}

async function cleanupTf(synthDir: string) {
  await withSpinner("terraform destroy", () => terraformDestroy(synthDir));
  rmSync(synthDir, { recursive: true, force: true });
}

async function isTerraformInstalled(synthDir: string) {
  const output = await execCapture("terraform version", { cwd: synthDir });
  return output.startsWith("Terraform v");
}

export async function terraformInit(synthDir: string) {
  return execCapture("terraform init", { cwd: synthDir });
}

async function terraformApply(synthDir: string) {
  return execCapture("terraform apply -auto-approve", { cwd: synthDir });
}

async function terraformDestroy(synthDir: string) {
  return execCapture("terraform destroy -auto-approve", { cwd: synthDir });
}

async function terraformOutput(synthDir: string, name: string) {
  const output = await execCapture("terraform output -json", { cwd: synthDir });
  const parsed = JSON.parse(output);
  if (!parsed[name]) {
    throw new Error(`terraform output ${name} not found`);
  }
  return parsed[name].value;
}

function pickOneTestPerEnvironment(testPaths: string[]) {
  // Given a list of test paths like so:
  //
  // root/Default/env0/a/b/test:test1
  // root/Default/env0/d/e/f/test:test2
  // root/Default/env0/g/test:test3
  // root/Default/env1/a/b/test:test1
  // root/Default/env1/d/e/f/test:test2
  // root/Default/env1/g/test:test3
  // root/Default/env2/a/b/test:test1
  // root/Default/env2/d/e/f/test:test2
  // root/Default/env2/g/test:test3
  //
  // This function returns a list of test paths which uses each test name
  // exactly once and each "env" exactly once. In this case, the result would
  // be:
  //
  // root/Default/env0/a/b/test:test1
  // root/Default/env1/d/e/f/test:test2
  // root/Default/env2/g/test:test3

  const tests = new Map<string, string>();
  const envs = new Set<string>();

  for (const testPath of testPaths) {
    const testSuffix = testPath.substring(testPath.indexOf("env") + 1); // "<env #>/<path to test>"
    const env = testSuffix.substring(0, testSuffix.indexOf("/")); // "<env #>"
    const test = testSuffix.substring(testSuffix.indexOf("/") + 1); // "<path to test>"

    if (envs.has(env)) {
      continue;
    }

    if (tests.has(test)) {
      continue;
    }

    tests.set(test, testPath);
    envs.add(env);
  }

  return Array.from(tests.values());
}

function sortTests(a: std.TestResult, b: std.TestResult) {
  if (a.pass && !b.pass) {
    return -1;
  }
  if (!a.pass && b.pass) {
    return 1;
  }
  return a.path.localeCompare(b.path);
}

const MAX_BUFFER = 10 * 1024 * 1024;

/**
 * Executes command and returns STDOUT. If the command fails (non-zero), throws an error.
 */
async function execCapture(command: string, options: { cwd: string }) {
  log(command);
  const exec = promisify(cp.exec);
  const { stdout, stderr } = await exec(command, {
    cwd: options.cwd,
    maxBuffer: MAX_BUFFER,
  });
  if (stderr) {
    throw new Error(stderr);
  }
  log(stdout);
  return stdout;
}
