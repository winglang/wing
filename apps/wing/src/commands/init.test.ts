import { readdir, writeFile } from "fs/promises";
import { join } from "path";
import inquirer from "inquirer";
import { describe, expect, test, vitest, beforeEach, afterEach, vi } from "vitest";
import { test as cliTest } from "./test/test";
import { init } from "../commands/init";
import { generateTmpDir, projectTemplateNames } from "../util";

vitest.mock("inquirer");

const templates = projectTemplateNames();

describe.each(templates)("new %s --language=wing", (template) => {
  let log: any;
  beforeEach(() => {
    log = console.log;
    console.log = vi.fn();
  });

  afterEach(() => {
    console.log = log;
  });

  test(`wing new ${template} && wing test main.w`, async () => {
    const workdir = await generateTmpDir();
    process.chdir(workdir);

    await init(template, { language: "wing" });

    await cliTest(["main.w"], {
      platform: ["sim"],
      clean: false,
    });
  });
});

describe("new --list-templates", () => {
  let log: any;
  beforeEach(() => {
    log = console.log;
    console.log = vi.fn();
  });

  afterEach(() => {
    console.log = log;
  });

  test("does not contain duplicate template names", async () => {
    await init("", { listTemplates: true });
    expect(console.log).toHaveBeenCalledWith(templates.join("\n"));

    const outputLines = (console.log as any).mock.calls[0][0].split("\n");
    const outputLinesAsSet = new Set(outputLines);

    expect(outputLines.length).toBeGreaterThan(0);
    // Check there are no duplicate lines
    expect(outputLinesAsSet.size).toBe(outputLines.length);
  });
});

describe("new empty --language=ts", () => {
  let log: any;
  beforeEach(() => {
    log = console.log;
    console.log = vi.fn();
  });

  afterEach(() => {
    console.log = log;
  });

  test(`wing new empty --language ts && wing test main.w`, async () => {
    const workdir = await generateTmpDir();
    process.chdir(workdir);

    await init("empty", { language: "typescript" });

    await cliTest(["main.ts"], {
      platform: ["sim"],
      clean: false,
    });
  });
});

describe("new", () => {
  let log: any;
  beforeEach(() => {
    log = console.log;
    console.log = vi.fn();
  });

  afterEach(() => {
    console.log = log;
  });

  test("wing new invalid-template", async () => {
    const workdir = await generateTmpDir();
    process.chdir(workdir);

    await expect(init("invalid-template")).rejects.toThrow(
      /Template "invalid-template" is not available/
    );
  });

  // test an invalid language
  test("wing new http-api --language=python", async () => {
    const workdir = await generateTmpDir();
    process.chdir(workdir);

    await expect(init("http-api", { language: "python" })).rejects.toThrow(/Unknown language/);
  });

  // test an unavailable language
  test("wing new http-api --language=typescript", async () => {
    const workdir = await generateTmpDir();
    process.chdir(workdir);

    await expect(init("http-api", { language: "typescript" })).rejects.toThrow(
      /Template "http-api" is not available in typescript/
    );
  });

  test("wing new with conflicting file", async () => {
    const workdir = await generateTmpDir();
    process.chdir(workdir);

    // write a conflicting file
    await writeFile(join(workdir, "main.w"), "hello");

    await expect(init("http-api", { language: "wing" })).rejects.toThrow(
      /The following files already exist in the current directory and will be overwritten/
    );
  });

  test("wing new with interactive prompt", async () => {
    const workdir = await generateTmpDir();
    process.chdir(workdir);

    (inquirer.prompt as any) = vitest.fn().mockResolvedValue({
      template: "http-api",
      language: "wing",
    });

    await init(undefined as any);

    const files = await readdir(workdir);
    expect(files).toEqual(["main.w", "package-lock.json", "package.json"]);
  });
});
