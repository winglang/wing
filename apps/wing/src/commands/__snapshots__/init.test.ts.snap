// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`projectTemplateNames > should list project templates 1`] = `
[
  "http-api",
]
`;
